### PR TITLE
Feat: hardware details tree filter

### LINKED ISSUE #499: Hardware Trees Header: Backend With Filter

### LINKED ISSUE #500: Hardware Trees Header: Frontend With Filter

### PR DESCRIPTION
**Pull request by MarceloRobert**
Enables filtering for trees in the Hardware Details page and selection for the header table.

## How to test:
Enable dev feature flag before running the front-end by running `export VITE_FEATURE_FLAG_SHOW_DEV=true`
Enter the hardware details page and try out the row selection, check the url to see how it is sent to the backend and you may try to change the selection manually by changing the url. 

PIGGYBACKING: fixes unnecessary and wrongfully passed search parameters when clicking on the breadcrumbs.

Closes #500 
Part of #446

### CODE REVIEW

File: dashboard/src/pages/hardwareDetails/HardwareDetails.tsx [hunk 1]
```diff
@@ -14,19 +16,55 @@ import {
...
+      });
+    },
+    [navigate],
+  );
+
   const { data, isLoading } = useHardwareDetails(
     hardwareId,
     intervalInDays,
     origin,
+    treeIndexes ?? [],
```
Comment (Francisco2002) [inline dashboard/src/pages/hardwareDetails/HardwareDetails.tsx:63]: When we enter in `HardwareDetails` for the first time, the `treeIndexes` is `undefined` and a empty list of trees ids is passed as params. But initially, all trees are selected. We should pass all the trees ids in this case.

File: dashboard/src/pages/hardwareDetails/HardwareDetailsHeaderTable.tsx [hunk 2]
```diff
@@ -23,35 +25,43 @@ import type { Trees } from '@/types/hardware/hardwareDetails';
...
-// import { IndeterminateCheckbox } from '@/components/Checkbox/IndeterminateCheckbox';
+import { IndeterminateCheckbox } from '@/components/Checkbox/IndeterminateCheckbox';
+import { useFeatureFlag } from '@/hooks/useFeatureFlag';
+
+const DEBOUNCE_INTERVAL = 1000;
 
 interface IHardwareHeader {
   treeItems: Trees[];
+  selectedIndexes?: number[];
+  updateTreeFilters: (selecteds?: number[]) => void;
```
Comment (Francisco2002) [inline dashboard/src/pages/hardwareDetails/HardwareDetailsHeaderTable.tsx:36]: nit: `selectedTreesId` is more suggestive.

File: dashboard/src/pages/hardwareDetails/HardwareDetailsHeaderTable.tsx [hunk 3]
```diff
@@ -98,46 +108,86 @@ const columns: ColumnDef<Trees>[] = [
...
   const [pagination, setPagination] = useState<PaginationState>({
     pageIndex: 0,
     pageSize: 5,
   });
-  // const [rowSelection, setRowSelection] = useState({});
 
-  const data = useMemo(() => {
-    return sanitizeTreeItems(treeItems);
-  }, [treeItems]);
+  const { showDev } = useFeatureFlag();
```
Comment (lfjnascimento) [inline dashboard/src/pages/hardwareDetails/HardwareDetailsHeaderTable.tsx:139]: remove the dev only feature

File: dashboard/src/pages/hardwareDetails/HardwareDetailsHeaderTable.tsx [hunk 4]
```diff
@@ -98,46 +106,80 @@ const columns: ColumnDef<Trees>[] = [
...
+          i.toString(),
+          true,
+        ]),
+      );
+    } else {
+      return Object.fromEntries(
+        Array.from(selectedIndexes, treeIndex => [treeIndex.toString(), true]),
+      );
+    }
+  }, [selectedIndexes, treeItems]);
```
Comment (lfjnascimento) [inline dashboard/src/pages/hardwareDetails/HardwareDetailsHeaderTable.tsx:150]: - create a function outside the component for this logic
- nit: when you use `if(..) return` you don't need the `else` just `return`

File: dashboard/src/pages/hardwareDetails/HardwareDetailsHeaderTable.tsx [hunk 5]
```diff
@@ -98,46 +106,80 @@ const columns: ColumnDef<Trees>[] = [
...
+      const updatedSelection = indexesFromRowSelection(
+        rowSelection,
+        treeItems.length,
+      );
+      updateTreeFilters(updatedSelection);
+    }, DEBOUNCE_INTERVAL);
+    return (): void => {
+      clearTimeout(handler);
+    };
+  }, [rowSelection, treeItems.length, updateTreeFilters]);
```
Comment (lfjnascimento) [inline dashboard/src/pages/hardwareDetails/HardwareDetailsHeaderTable.tsx:165]:
```suggestion
  const rowSelectionDebounced = useDebounce(rowSelection, DEBOUNCE_INTERVAL);

  useEffect(() => {
    const updatedSelection = indexesFromRowSelection(
      rowSelectionDebounced,
      treeItems.length,
    );
    updateTreeFilters(updatedSelection);
  }, [rowSelectionDebounced, treeItems.length, updateTreeFilters]);
```

File: dashboard/src/routes/hardware/$hardwareId/route.tsx [hunk 6]
```diff
@@ -6,6 +6,7 @@ import { zPossibleValidator, zTableFilterInfo } from '@/types/tree/TreeDetails';
 
 const hardwareDetailsSearchSchema = z.object({
   currentPageTab: zPossibleValidator,
+  treeFilter: z.array(z.number().int()).optional(),
```
Comment (Francisco2002) [inline dashboard/src/routes/hardware/$hardwareId/route.tsx:9]: nit: `treeIndexesList` or `treeSelecteds` are more suggestive

File: dashboard/src/pages/hardwareDetails/HardwareDetails.tsx [hunk 7]
```diff
@@ -14,19 +16,52 @@ import {
...
+    treeName: tree['treeName'] ?? '-',
+    gitRepositoryBranch: tree['gitRepositoryBranch'] ?? '-',
+    headGitCommitName: tree['headGitCommitName'] ?? '-',
+    headGitCommitHash: tree['headGitCommitHash'] ?? '-',
+    gitRepositoryUrl: tree['gitRepositoryUrl'] ?? '-',
+    index: tree['index'],
+  }));
+
 function HardwareDetails(): JSX.Element {
   const searchParams = useSearch({ from: '/hardware/$hardwareId' });
```
Comment (Francisco2002) [inline dashboard/src/pages/hardwareDetails/HardwareDetails.tsx:35]: use destructuring to get the fields used only.

File: dashboard/src/routes/hardware/$hardwareId/route.tsx (hunk 6 above)
Comment (MarceloRobert) [reply dashboard/src/routes/hardware/$hardwareId/route.tsx:9]: I'll use `treeIndexes` so to not make it too long or too illogical. I'm also renaming it for that name in hardwareDetails already anyway.

**Review by lfjnascimento — APPROVED** [review]
looks nice, pre-approving

File: dashboard/src/pages/hardwareDetails/HardwareDetailsHeaderTable.tsx [hunk 8]
```diff
@@ -98,46 +107,78 @@ const columns: ColumnDef<Trees>[] = [
...
+  rowSelection: RowSelectionState,
+  maxTreeItems: number,
+): number[] => {
+  const rowSelectionValues = Object.values(rowSelection);
+  if (
+    rowSelectionValues.length === maxTreeItems ||
+    rowSelectionValues.length === 0
+  ) {
+    return [];
+  } else {
```
Comment (Francisco2002) [inline dashboard/src/pages/hardwareDetails/HardwareDetailsHeaderTable.tsx:134]: You can remove this else too

**Review by Francisco2002 — APPROVED** [review]
nice!!
pre approving

**Pull request merged**